payments/views.py

```diff
@@ -98,129 +101,97 @@ def setup_class(cls) -> SubscriptionInformation:
-            return None
+        customer_id = membership.customer
 
-        # Attach the payment source to the customer
-        _, err = form.attach_to_customer(membership.customer)
+        # Build return URL to redirect back after portal session
+        return_url = self.request.build_absolute_uri(reverse("setup_subscription"))
 
-        # if the error is not, return
-        if err is not None:
+        # Determine the price ID dynamically based on the tier the user selected
+        tier = membership.tier
+        price_id_map = {
+            TierField.CONTRIBUTOR: settings.STRIPE_CONTRIBUTOR_PRICE_ID,
+            TierField.PATRON: settings.STRIPE_PATRON_PRICE_ID,
+            TierField.STARTER: settings.STRIPE_STARTER_PRICE_ID,
+        }
+        price_id = price_id_map.get(tier)
+
+        if not price_id:
             form.add_error(
                 None,
-                "Something went wrong when talking to our payment service provider. Please try again later or contact support. ",
+                "Invalid membership tier selected. Please contact support. ",
             )
             return None
 
-        # we went to the starter tier
-        if not form.user_go_to_starter:
-            instance = membership.create_subscription()
-        else:
-            instance = membership.cancel_create_subscription()
+        # Create the portal session via the Stripe API directly
+        stripe.api_key = settings.STRIPE_SECRET_KEY
+
+        try:
+            session = stripe.checkout.Session.create(
+                customer=customer_id,
+                payment_method_types=["card"],
+                mode="subscription",
+                line_items=[
+                    {
+                        "price": price_id,
+                        "quantity": 1,
+                    }
+                ],
+                success_url=return_url,
+                cancel_url=return_url,
+            )
+            portal_url = session.url
+            err = None
+        except Exception as e:
+            portal_url = None
+            err = str(e)
 
-        if instance is None:
+        if err is not None or not portal_url:
             form.add_error(
                 None,
-                "Something went wrong trying to create the subscription. Please try again later or contact support. ",
+                "Something went wrong when creating the checkout session. Please try again later or contact support. ",
             )
+            return None
 
-        return instance
-
-    def dispatch_success(self, validated: SubscriptionInformation) -> HttpResponse:
-        """called upon successful setup"""
+        # Redirect the user to the Stripe portal session
+        return portal_url
 
-        # if this was not created from an update operation, do nothing
-        if not validated.created_from_update:
-            return super().dispatch_success(validated)
+    def dispatch_success(self, validated: Any) -> HttpResponse:
```

Comment:
If you're passing validated to the redirect response, shouldn't the type be str or something?